public/app/plugins/datasource/prometheus/datasource.tsx

```diff
@@ -451,6 +451,34 @@ export class PrometheusDatasource
     return processedTargets;
   }
 
+  // @ts-ignore
+  transform(
+    request: DataQueryRequest<PromQuery>,
+    response: Observable<DataQueryResponse>
+  ): Observable<DataQueryResponse> {
+    let requestInfo: CacheRequestInfo | undefined = undefined;
+    if (this.hasIncrementalQuery) {
+      requestInfo = this.cache.requestInfo(request, this.interpolateString.bind(this));
+    }
+
+    const startTime = new Date();
+
+    return response.pipe(
+      map((response) => {
+        const amendedResponse = {
+          ...response,
+          data: this.cache.procFrames(request, requestInfo, response.data),
+        };
+        return transformV2(amendedResponse, request, {
+          exemplarTraceIdDestinations: this.exemplarTraceIdDestinations,
+        });
+      }),
+      tap((response: DataQueryResponse) => {
+        trackQuery(response, request, startTime);
+      })
```

Comment:
Is there a good reason why this transformation is in the frontend and not the backend?

Reply:
i think the simple explanation is that no one ported it Go :sweat_smile: 

i opened this a while back hoping someone would pick it up: https://github.com/grafana/grafana/issues/46608.

that being said, a user (@quentinmit) has recently pointed out that moving this to the backend (instead of even into the panels) would necessitate all datasources capable of producing prometheus-like histograms to (re)implement this de-accumulation step:

https://github.com/grafana/grafana/issues/46608#issuecomment-1636609252
https://github.com/grafana/grafana/issues/69018#issuecomment-1636610096

> +1 to this question. Think Prometheus implements some sort of query caching on the client side, @leeoniya should have some more details about it.

we do have incremental query caching on the client side, but we should be able to do this cache post-transform. i think @gtk-grafana and i worked through the issue mentioned in https://github.com/grafana/grafana/issues/69807.

Reply:
Yeah, post-transform query cache was working fine for me locally, so I don't think that should hold up migrating the transformv2 to the backend.